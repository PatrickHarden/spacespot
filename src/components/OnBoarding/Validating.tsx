import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { useIntl } from 'react-intl'
import styled from 'styled-components'

import actions from 'services/onboarding/actions'
import onboardingSelectors from 'services/onboarding/selectors'
import BuildingSeparator from './BuildingSeparator'
import Spinner from 'components/icons/Spinner'
import history from 'browserhistory'
import FilledButton from 'components/common/FilledButton'
import Heading1 from 'components/common/Heading1'
import Heading2 from 'components/common/Heading2'
import Container from './MainContainer'

import hourglass from 'assets/icons/hourglass.svg'

const EditedContainer = styled.div`
  text-align: center;
`
const Title = styled(Heading1)``

const SpinnerContainer = styled.div`
  text-align: center;
  margin-top: 20px;
`
const Wait5MinsIcon = styled.img`
  padding: 70px 0 10px 0;
`

const EditedTitle = styled(Heading2)`
  margin-bottom: 0;
`

const EditedText = styled(Heading2)`
  margin-top: 0;
`

const ListSpaces = styled(FilledButton)`
  margin-top: 40px;
  width: 176px;
  background-color: #404042;
  color: #ffffff;
`

const Validating = () => {
  const dispatch = useDispatch()
  const isPublshing = useSelector(onboardingSelectors.onboardingIsPublishing)
  const { formatMessage } = useIntl()
  const t = (s: string) => formatMessage({ id: s })
  return (
    <Container>
      {isPublshing ? (
        <>
          <Title>{t('ONBOARDING_TITLE')}</Title>
          <BuildingSeparator />
          <SpinnerContainer>
            <Spinner />
          </SpinnerContainer>
        </>
      ) : (
        <EditedContainer>
          <Wait5MinsIcon src={hourglass} alt="hourglass" />
          <EditedTitle>{t('ONBOARDING_EDITED_TITLE')}</EditedTitle>
          <EditedText>{t('ONBOARDING_EDITED_TEXT')}</EditedText>
          <ListSpaces
            onClick={() => {
              dispatch(actions.resetSpaces())
              history.push(`/onboarding/`)
            }}>
            {t('ONBOARDING_EDITED_LIST_SPACES')}
          </ListSpaces>
        </EditedContainer>
      )}
    </Container>
  )
}
export default Validating
