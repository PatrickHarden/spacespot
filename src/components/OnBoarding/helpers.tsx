import {
  OnBoardingState,
  SpaceData,
  SpaceStatus,
  AvailabilityFlexPriceInterval,
} from 'services/onboarding/types'

const date = new Date('2020/01/01')

export const createSpace = (type: string): SpaceData => ({
  id: '0',
  type,
  spaceName: '',
  spaceDescription: '',
  spaceHighlights: '',
  spaceSize: undefined,
  spaceFloor: undefined,
  spaceSizeCommon: undefined,
  spaceMatterPort: '',
  spaceFloored: '',
  use: '',
  spaceRent: undefined,
  spaceServices: undefined,
  spaceServicesNotNegotiable: false,
  availabilityFixed: date,
  months: undefined,
  hotDesks: {
    desks: 0,
    availableFrom: date,
    minLease: 0,
    price: 0,
    currencyCode: 'NOK',
    frequency: AvailabilityFlexPriceInterval.Monthly,
  },
  fixedDesks: {
    desks: 0,
    availableFrom: date,
    minLease: 0,
    price: 0,
    currencyCode: 'NOK',
    frequency: AvailabilityFlexPriceInterval.Monthly,
  },
  servicedOffices: [],
  images: [],
  floorPlan: [],
  fitout: [{ description: 'FitOut1', name: 'FitOut1', amount: 1000 }],
  status: SpaceStatus.ReadyForPublish,
})

export const createOnBoardingState = (): OnBoardingState => ({
  isNew: true,
  buildingId: undefined,
  images: [
    {
      key: 0,
      name: '40VilliersStreet.External_Photo_1_small.png',
      type: 'image/png',
      uri:
        'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8/5+hHgAHggJ/PchI7wAAAABJRU5ErkJggg==',
    },
  ],
  docs: [],
  name: '',
  description: '',
  address: '',
  city: '',
  postCode: '',
  country: '',
  location: '',
  latLng: undefined,
  district: '',
  subdistrict: {
    id: 0,
    name: '',
  },
  amenities: { hasGym: { checked: true, desc: 'Gym' } },
  spaces: { '0': createSpace('FLEX') },
  spaceCount: 1,
})

export const valInteger = (
  current: number | undefined,
  value: number,
  min: number,
  max?: number,
): number => {
  const val = Number(Number(value).toFixed())
  return val < min || (max && val > max) ? current || 0 : val
}
