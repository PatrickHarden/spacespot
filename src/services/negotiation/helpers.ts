import { get } from 'lodash'
import {
  NegotiationTerms,
  NegotiationTerm,
  NegotiationTermFields,
  NegotiationTermFieldsIds,
} from './types'

export const getNegotiationTermArray = (
  terms: NegotiationTerms,
  ids: NegotiationTermFieldsIds,
) => {
  const values: Array<NegotiationTerm> = Object.keys(terms).map(key => {
    return {
      termId: ids[key],
      value: get(terms, key),
    }
  })
  return values
}

const decapitalize = (title: string) => title[0].toLowerCase() + title.slice(1)

export const getFields = (
  fields: NegotiationTermFields,
  negTerms: Array<NegotiationTerm>,
): NegotiationTerms => {
  const initialTerms = {} as NegotiationTerms
  if (!negTerms || !fields) return initialTerms
  return (negTerms.reduce(
    (acc, term) => ({
      ...acc,
      [decapitalize(fields[term.termId])]: term.value,
    }),
    initialTerms,
  ) as unknown) as NegotiationTerms
}
