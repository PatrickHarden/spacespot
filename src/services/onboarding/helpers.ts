import { get, cloneDeep } from 'lodash'
import { Space } from 'services/space/types'
import spaceSelectors from 'services/space/selectors'
import {
  AmenityOptions,
  OnBoardingState,
  AvailabilityFlexPriceInterval,
  AvailabilityFlex,
} from 'services/onboarding/types'

export const aspects2Amenities = (
  amenities: AmenityOptions,
  aspects: Array<string>,
): AmenityOptions => {
  const newAmenities = cloneDeep(amenities) || {}
  if (!aspects) {
    return newAmenities
  }
  aspects.forEach(key => {
    if (newAmenities[key]) {
      newAmenities[key].checked = true
    } else {
      newAmenities[key] = {
        checked: true,
        desc: key,
      }
    }
  })
  return newAmenities
}

export const loadOnboardingSpace = (
  defaultAmenities: AmenityOptions | undefined,
  space: Space,
  isBuilding: boolean,
): OnBoardingState => {
  const building = spaceSelectors.translateBuilding(space, isBuilding)
  return {
    buildingId: building.buildingId,
    isNew: false,
    images: building.images || [],
    docs: [],
    name: building.name,
    description: building.description,
    address: building.address,
    city: building.city,
    postCode: building.postCode,
    country: building.country,
    location: building.location || '',
    latLng: building.latLng,
    district: building.district,
    subdistrict: building.subdistrict,
    amenities: aspects2Amenities(
      defaultAmenities || {},
      get(space, 'Common.Aspects', []),
    ),
    spaces: isBuilding ? {} : spaceSelectors.translateSpace(space),
    spaceCount: 0,
    uploadedImages: building.uploadedImages,
    uploadedDocs: building.uploadedDocs,
    isPublishing: false,
    defaultAmenities: defaultAmenities,
  }
}

const defaultDate = new Date()
defaultDate.setDate(new Date().getDate() + 1)

export const getDefaultAvailability = (): AvailabilityFlex => ({
  desks: 0,
  availableFrom: defaultDate,
  minLease: 0,
  price: 0,
  currencyCode: 'NOK',
  frequency: AvailabilityFlexPriceInterval.Monthly,
})
