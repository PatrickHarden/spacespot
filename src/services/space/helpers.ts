import { get, chain } from 'lodash'
import { Space } from './types'
import {
  CommonChargeType,
  CommonImage,
  ImageResource,
  CommonPostalAddress,
  CommonActualAddress,
} from 'services/space/PropertyListingschema'
import { FLEX_TYPE_CONSTANT, FIXED_TYPE_CONSTANT } from '../onboarding/utils'
import { routingLang } from 'intl'

import CONFIG from 'config'

export const generateGallery = (item: any) => {
  const [url] = item['Common.ImageResources'].filter(
    (photo: any) => photo['Common.Breakpoint'] === 'small',
  )
  const alt = item['Common.ImageCaption']
  return {
    url: url['Source.Uri'],
    alt,
  }
}

type Charges = {
  rent: number
  serviceCharges: number
  additionalCosts: number
  currencyCode: string
  total: number
}

export const getCharges = (space: Space) => {
  let charges: Charges = {
    rent: 0,
    serviceCharges: 0,
    additionalCosts: 0,
    currencyCode: 'EUR',
    total: 0,
  }
  const ccharges: Array<CommonChargeType> = get(space, 'Common.Charges', [])
  charges = ccharges.reduce((acc, charge) => {
    const kind = charge['Common.ChargeKind']
    const value = charge['Common.Amount'] || 0
    const currency = charge['Common.CurrencyCode']
    if (currency) {
      acc.currencyCode = currency
    }
    switch (kind) {
      case 'Rent':
        acc.rent = value
        break
      case 'ServiceCharge':
        acc.serviceCharges += value
        break
      default:
        acc.additionalCosts += value
    }
    acc.total += value
    return acc
  }, charges)
  return charges
}

export const getImages = (space: Space) => {
  const photos = get(space, 'Common.Photos')
  const urls = photos
    ? photos.map((item: CommonImage) => {
      const [original] = item['Common.ImageResources'].filter(
        (photo: ImageResource) => photo['Common.Breakpoint'] === 'original',
      )
      const [small] = item['Common.ImageResources'].filter(
        (photo: ImageResource) => photo['Common.Breakpoint'] === 'small',
      )
      return {
        original:
          original['Source.Uri'] ||
          CONFIG.GL_PHOTO_HOST + original['Common.Resource.Uri'],
        thumbnail:
          small['Source.Uri'] ||
          CONFIG.GL_PHOTO_HOST + small['Common.Resource.Uri'],
      }
    })
    : []
  return urls
}

export const getFloorPlans = (space: Space) => {
  const floorPlans = get(space, 'Common.FloorPlans')
  const urls = floorPlans
    ? floorPlans.map((item: CommonImage) => {
      const original = item['Common.ImageResources'][0]
      return {
        original:
          original['Source.Uri'] ||
          CONFIG.GL_PHOTO_HOST + original['Common.Resource.Uri'],
        thumbnail:
          original['Source.Uri'] ||
          CONFIG.GL_PHOTO_HOST + original['Common.Resource.Uri'],
      }
    })
    : []
  return urls
}

export const getPhoto = (space: Space): string => {
  const photo = chain(space)
    .get('Common.Photos')
    .head()
    .get('Common.ImageResources')
    .filter((img: ImageResource) => img['Common.Breakpoint'] === 'small')
    .head()
    .get('Source.Uri')
    .value()
  return photo || ''
}

export const getType = (space: Space) =>
  get(space, 'Common.UsageType') === 'Office' //FIXME when mapping changes
    ? FIXED_TYPE_CONSTANT
    : FLEX_TYPE_CONSTANT

export const isFixed = (space: Space) => getType(space) === FIXED_TYPE_CONSTANT

export const getSize = (space: Space) => {
  if (isFixed(space)) {
    return get(space, "['Common.MinimumSize'][0]['Common.Size']", 0)
  }
  return 0 // FIXME
}

/**
 * Size units (i.e: sqm)
 */
export const getSizeUnits = (space: Space) => {
  if (isFixed(space)) {
    return get(space, "['Common.MinimumSize'][0]['Common.Units']", 0)
  }
  return 0 // FIXME
}

/**
 * Calculate desks from size for fixed spaces
 *
 * desks = size in sqm / 15
 *
 * - For flex/coworking returns empty string
 * - If units != sqm or sqft, returns size with units
 *
 */
export const getDesks = (space: Space) => {
  if (isFixed(space)) {
    const sz: number = getSize(space)
    const units = getSizeUnits(space)
    if (units === 'sqm' || units === 'm2') {
      return (sz / 15).toFixed()
    } else if (units === 'sqft') {
      return ((sz * 0.092903) / 15).toFixed()
    } else {
      // unknown units
      return `${sz} ${units}`
    }
  }
  return ''
}

/**
 * Get common area (shared) size
 *
 * Shared common area = TotalSize - Size
 */
export const getSizeCommon = (space: Space) => {
  if (isFixed(space)) {
    return (
      get(space, "['Common.TotalSize'][0]['Common.Size']", 0) - getSize(space)
    )
  }
  return 0
}

export const getFloor = (space: Space) => get(space, 'Common.FloorNumber')

const getPostalAddress = (space: Space): CommonPostalAddress => {
  const actual: CommonActualAddress = get(space, 'Common.ActualAddress')
  const postal: Array<CommonPostalAddress> = get(
    actual,
    'Common.PostalAddresses',
    [],
  )
  return postal[0]
}

export const getLocation = (space: Space) =>
  get(space, '["Common.LocationDescription"][0]["Common.Text"]')

export const getTitle = (space: Space): string =>
  get(
    space,
    '["Common.Strapline"][0]["Common.Text"]',
    get(getPostalAddress(space), 'Common.Line1', ''),
  )

export const getStreet = (space: Space): string => {
  const address = getPostalAddress(space)
  return get(address, 'Common.Line2', get(address, 'Common.Line1', ''))
}

export const getDistrict = (space: Space) => {
  const address = getPostalAddress(space)
  return get(address, 'Common.Line3', '')
}

export const getSubdistrict = (space: Space) => {
  const address = getPostalAddress(space)
  return {
    id: 0,
    name: get(address, 'Common.Line4', ''),
  }
}

export const getCity = (space: Space): string => {
  const address = getPostalAddress(space)
  return get(address, 'Common.Locallity', '')
}

export const getPK = (space: Space): string =>
  get(space, 'Common.PrimaryKey', '')

export const getSpaceURI = (space: Space): string => {
  const office = routingLang({ nb: 'kontor', default: 'office' })
  const coworking = 'coworking'

  const id = getPK(space)
  const address = getStreet(space).replace(/\//g, '-')
  const district = (getDistrict(space) || 'NA').replace(/\//g, '-')
  const subdistrict = (getSubdistrict(space).name || 'NA').replace(/\//g, '-')
  const city = getCity(space)
  const title = getTitle(space).replace(/\//g, '-')
  const stype = isFixed(space) ? office : coworking
  const uri = `detail/${stype}/${city}/${district}/${subdistrict}/${id}/${title}-${address}`
  const repKeys: { [key: string]: string } = {
    æ: 'ae',
    å: 'aa',
    ø: 'oe',
    Æ: 'Ae',
    Å: 'Aa',
    Ø: 'Oe',
    á: 'a',
    é: 'e',
    í: 'i',
    ó: 'o',
    ú: 'u',
    Á: 'A',
    É: 'E',
    Í: 'I',
    Ó: 'O',
    Ú: 'U',
    '–': '-',
    '|': '-',
    ä: 'a',
    ë: 'e',
    ï: 'i',
    ö: 'o',
    ü: 'u',
    Ä: 'A',
    Ë: 'E',
    Ï: 'I',
    Ö: 'O',
    Ü: 'U',
  }
  return uri
    .replace(/['()&?,. ]/g, '-')
    .replace(/\*/g, '-')
    .replace(/[æåøÆÅØáéíóúÁÉÍÓÚ–|äëïöüÄËÏÖÜ]/g, char => repKeys[char] || '')
    .replace(/-+/g, '-')
}
