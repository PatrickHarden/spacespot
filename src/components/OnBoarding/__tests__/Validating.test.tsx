import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

import Validating from '../Validating'
import { createOnBoardingState } from '../helpers'

const mockState = createOnBoardingState()

jest.mock('react-redux', () => ({
  useSelector: () => () => mockState,
  useDispatch: () => {
    return jest.fn()
  },
}))

jest.mock('react-intl', () => ({
  useIntl: () => ({ formatMessage: (s: any) => s.id }),
}))

it('renders without crashing ', async () => {
  const wrapper = shallow(<Validating />)
  expect(shallowToJson(wrapper)).toMatchSnapshot()
})
