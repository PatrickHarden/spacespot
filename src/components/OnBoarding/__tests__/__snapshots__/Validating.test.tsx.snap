// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing  1`] = `
<styled.div>
  <Styled(styled.h1)>
    ONBOARDING_TITLE
  </Styled(styled.h1)>
  <BuildingSeparator />
  <styled.div>
    <Styled(Component) />
  </styled.div>
</styled.div>
`;
